pkg/mimir/runtime_config.go

```diff
@@ -50,6 +51,30 @@ func newTenantLimits(manager *runtimeconfig.Manager) validation.TenantLimits {
 	}
 }
 
+// AddChangeListener implements validation.TenantLimits.
+func (l *runtimeConfigTenantLimits) AddChangeListener(listener validation.ChangeListener) validation.ChangeListenerRef {
+	ch := l.manager.CreateListenerChannel(1)
+	lastConfig := l.manager.GetConfig()
+
+	go func() {
+		for range ch {
+			if currConfig := l.manager.GetConfig(); !reflect.DeepEqual(lastConfig, currConfig) {
+				listener()
```

Comment:
I'm not sure about it, from a design perspective, because of possible timing issues. I think this kind of listeners should just tell you something has changed, and then you lookup by yourself.